File: 382.linked-list-random-node.py
#
# @lc app=leetcode id=382 lang=python3
#
# [382] Linked List Random Node
#

# Time:  O()
# Space: O()

import random


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


# @lc code=start
class Solution:
    def __init__(self, head: ListNode):
        self.head = head

    def getRandom(self) -> int:
        res, cnt, curr = -1, 1, self.head
        while curr:
            if random.randint(1, cnt) == 1:
                res = curr.val
            cnt, curr = cnt + 1, curr.next
        return res


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()
# @lc code=end

File: 382.linked-list-random-node.js
/*
 * @lc app=leetcode id=382 lang=javascript
 *
 * [382] Linked List Random Node
 */

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

// @lc code=start
/**
 * @param head The linked list's head.
        Note that the head is guaranteed to be not null, so it contains at least one node.
 * @param {ListNode} head
 */
var Solution = function (head) {
  this.head = head;
};

/**
 * Returns a random node's value.
 * @return {number}
 */
Solution.prototype.getRandom = function () {
  let res = -1;
  let cnt = 1;
  let curr = this.head;
  while (curr) {
    if (Math.trunc(Math.random() * cnt) === 0) {
      res = curr.val;
    }
    ++cnt;
    curr = curr.next;
  }
  return res;
};

/**
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(head)
 * var param_1 = obj.getRandom()
 */
// @lc code=end
